### PR TITLE
[ADD] google_address_autocomplete, *: added autocomplete on address field

### PR DESCRIPTION
**Pull request by dhba-odoo**
*website_sale, website_sale_autocomplete
Purpose:
Writing addresses manually is error-prone, and a wrong address could lead to disappointments in case of on-site meetings, deliveries, snail mail, etc. This feature aims to provide accurate address suggestions based on user input to enhance the reliability of address entries in various modules such as res.company, res.partner, and crm.lead.

Specifications:

- Introduced a new widget google_address_autocomplete compatible with char fields.
- Integrated new widget with Google Places API to fetch address suggestions dynamically based on user input.
- Automatically populates additional address fields such as zip_field, city_field, and country_field when a user selects an address suggestion from the dropdown.
- Made the widget and its options available in Studio.
- The widget provides real-time address suggestions as the user types and automatically fills in the specified address fields upon selection.

Task-3851045

Enterprise counterpart PR: https://github.com/odoo/enterprise/pull/68201

### CODE REVIEW

**Review by kebeclibre — CHANGES_REQUESTED** [review]

File: addons/address_autocomplete/static/src/js/address_autocomplete.js
Comment (aab-odoo) [reply addons/address_autocomplete/static/src/js/address_autocomplete.js:61]: > @aab-odoo In theory, do you think that a widget could update multiple other fields's data at the same time ?

Yes, it is supported

File: addons/address_autocomplete/models/res_partner.py
Comment (aab-odoo) [inline addons/address_autocomplete/models/res_partner.py:15]: Not a fan of doing this in py instead of xml. At the very least, I would only apply the widget if no widget is set on the node (i.e. fallback, instead of forcing that widget)

File: addons/address_autocomplete/static/src/xml/extended_autocomplete_template.xml
Comment (dhba-odoo) [reply addons/address_autocomplete/static/src/xml/extended_autocomplete_template.xml]: I want an image to be displayed below the dropdown, similar to how it works on the website_sale_autocomplete. I tried achieving this without extending the autocomplete, but it’s not working that is why i need to extend it.
I think the issue might be related to `<t-esc>` that is used in autocomplete?

File: addons/address_autocomplete/static/src/js/address_autocomplete.js
Comment (dhba-odoo) [reply addons/address_autocomplete/static/src/js/address_autocomplete.js]: Yes, we can implement this in python, but in this case, we need to add it in the controller. Afterward, we can use an RPC call to retrieve the value.

File: addons/address_autocomplete/static/src/js/address_autocomplete.js
Comment (dhba-odoo) [reply addons/address_autocomplete/static/src/js/address_autocomplete.js]: > wrongly accessing fields: it should be accessed via record, not model.

Done

> If it is not generic and only applies to res.partner-like models, then we should have a way to prevent its use in other contexts.

Yes, this applies to modules like res_partner, res_company,etc.

> this field seems very coupled to how the model is declared. What if a model wants to use that, but zip is called "postal_code" ?

I need to update the address_fields when selecting options from the dropdown. I noticed that in res_partner, the address_fields are declared in the same way. As you mentioned, field names can sometimes differ. In this case, how can we identify whether a field is related to an address or not?
https://github.com/odoo/odoo/blob/b3d78b5b5c28a09ebfd6a6b4f5130b312fe0821c/odoo/addons/base/models/res_partner.py#L37

File: addons/address_autocomplete/__manifest__.py
Comment (dhba-odoo) [reply addons/address_autocomplete/__manifest__.py]: > have res.config setting to do so.

Now, since both the website_sale_autocomplete and the address_autocomplete module use the same api key google_places_api_key and res.config.settings, so which module should I install from res.config.settings? Should I install the website_sale_autocomplete, or the address_autocomplete.

**Comment by dhba-odoo** [comment]
@kebeclibre  @aab-odoo Waiting for your response.

**Comment by kebeclibre** [comment]
DIFF Explainer:
tests: + ~550 lines
business: + ~200 lines

**Comment by kebeclibre** [comment]
@JKE-be Here we go, modulo some fixes to come in function of functional testing, this PR should be nearly ready.
Tell me what you think about the website_sale_autocomplete thing

@odoo/rd-security Can I have a green flag here ? the requests.get has just been moved from another module

**Comment by kebeclibre** [comment]
> > the requests.get has just been moved from another module
> 
> That is not a justification, there's tons of old code that's broken _and_ the safety of a piece of code are often a function of their context or usage.

ai'ght:
PY006
1. requested URLs should be hard-coded or carefully constructed / checked to avoid being an open relay or giving the ability to access DMZ content 
CHECKED: The protocal/server part is hardcoded. The route is not, but is hardcoded in the caller which is a private method. None of the url parts is dynamic as a function of user input. Search parameters are, of course.

2. requests must have a timeout to mitigate slow network or slowloris-type attacks
CHECKED

3. consider using sessions if doing http requests in loops (sessions enable keep-alive), do not store sessions on models & al as they are not thread-safe and would allow direct unchecked use
NOT APPLICABLE

**Review by JKE-be — APPROVED** [review]
LGTM with a few comments based on our DM discussion:

- Add a parameter for debounce to reduce the number of requests.
- Include a parameter for minimal length, aligned with ICP requirements.
- Avoid triggering searches if the last search result is identical (e.g., on focus, select, or click events in the field).
- Use Assert for internal checks with is_internal().
- Clarify internal naming conventions—"internal" currently means backend concepts; consider a more explicit term?
- Add autocomplete="new-password" to the field to disable Chrome's autocomplete interfering with Google Place suggestions.
- Enable navigation of the selected record via keyboard by leveraging the aria-activedescendant attribute.
- Handle exceptions when "Enter" is pressed without a selection (avoid traceback). Side effect of not seeing what's selected.
- 'sublocality_level_1': ['sublocality_level_1'], which does not currently map to any Odoo field?

**Comment by JKE-be** [comment]
:+1: 
Thanks for all your changes.

https://github.com/odoo/odoo/pull/176262/files#diff-a4bc64c6e8e583db81549c056532df45b599103468a2555e9f68cbef3fb27297R15

This doesn’t match the ICP limit, so we continue to have RPC calls locally, but not to Google.
It's already a good improvement. However, there is a side effect: if ICP < 5, the JS will ignore it, which indicates a misconfiguration... not worst that a extra rcp in js to know this limit...


About the select with keyboard, I'm agree, it is already like that even in front end...
A nice imp to be done in a future task imo.


Lastly, you need to cast the ICP value of minimal_partial_address_size to an integer, as it is currently a string.
https://github.com/odoo/odoo/pull/176262/files#diff-42b1961362de68a4b1101d3ef3b8ac075442618de63622e94cc4d644268a7e43R81
```py
if len(partial_address) <= minimal_input_size:
    TypeError: '<=' not supported between instances of 'int' and 'str'
   ```


Thank you for your attention to the website_sale_* brother modules, your PR, and the quick follow-up.


LGTM to be merged after the ICP fix.

**Comment by kebeclibre** [comment]
robodoo rebase-ff r+

**Pull request closed without merging**
